src/wallet/asyncrpcoperation_shieldcoinbase.cpp

```diff
@@ -0,0 +1,441 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "asyncrpcqueue.h"
+#include "amount.h"
+#include "core_io.h"
+#include "init.h"
+#include "main.h"
+#include "net.h"
+#include "netbase.h"
+#include "rpcserver.h"
+#include "timedata.h"
+#include "util.h"
+#include "utilmoneystr.h"
+#include "wallet.h"
+#include "walletdb.h"
+#include "script/interpreter.h"
+#include "utiltime.h"
+#include "rpcprotocol.h"
+#include "zcash/IncrementalMerkleTree.hpp"
+#include "sodium.h"
+#include "miner.h"
+
+#include <iostream>
+#include <chrono>
+#include <thread>
+#include <string>
+
+#include "asyncrpcoperation_shieldcoinbase.h"
+
+using namespace libzcash;
+
+static int find_output(UniValue obj, int n) {
+    UniValue outputMapValue = find_value(obj, "outputmap");
+    if (!outputMapValue.isArray()) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Missing outputmap for JoinSplit operation");
+    }
+
+    UniValue outputMap = outputMapValue.get_array();
+    assert(outputMap.size() == ZC_NUM_JS_OUTPUTS);
+    for (size_t i = 0; i < outputMap.size(); i++) {
+        if (outputMap[i].get_int() == n) {
+            return i;
+        }
+    }
+
+    throw std::logic_error("n is not present in outputmap");
+}
+
+AsyncRPCOperation_shieldcoinbase::AsyncRPCOperation_shieldcoinbase(
+        std::vector<ShieldCoinbaseUTXO> inputs,
+        std::string toAddress,
+        CAmount fee,
+        UniValue contextInfo) :
+        inputs_(inputs), fee_(fee), contextinfo_(contextInfo)
+{
+    if (fee < 0 || fee > MAX_MONEY) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Fee is out of range");
+    }
+
+    if (inputs.size() == 0) {
+        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Empty inputs");
+    }
+
+    //  Check the destination address is valid for this network i.e. not testnet being used on mainnet
+    CZCPaymentAddress address(toAddress);
+    try {
+        tozaddr_ = address.Get();
+    } catch (const std::runtime_error& e) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, string("runtime error: ") + e.what());
+    }
+
+    // Log the context info
+    if (LogAcceptCategory("zrpcunsafe")) {
+        LogPrint("zrpcunsafe", "%s: z_shieldcoinbase initialized (context=%s)\n", getId(), contextInfo.write());
+    } else {
+        LogPrint("zrpc", "%s: z_shieldcoinbase initialized\n", getId());
+    }
+
+    // Lock UTXOs
+    lock_utxos();
+}
+
+AsyncRPCOperation_shieldcoinbase::~AsyncRPCOperation_shieldcoinbase() {
+}
+
+void AsyncRPCOperation_shieldcoinbase::main() {
+    if (isCancelled()) {
+        unlock_utxos(); // clean up
+        return;
+    }
+
+    set_state(OperationStatus::EXECUTING);
+    start_execution_clock();
+
+    bool success = false;
+
+#ifdef ENABLE_MINING
+  #ifdef ENABLE_WALLET
+    GenerateBitcoins(false, NULL, 0);
+  #else
+    GenerateBitcoins(false, 0);
+  #endif
+#endif
```

Comment:
This (and the reenablement) seems like it should be factored out if it is intended to apply to all async operations.